core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -436,6 +439,12 @@ public void commit() {
     }
   }
 
+  private void cleanupIfPossible(RuntimeException exception) {
+    if (!strictCleanup || exception instanceof CleanableFailure) {
+      Exceptions.suppressAndThrow(exception, this::cleanAll);
+    }
+  }
```

Comment:
I did a separate helper to avoid checkstyle complaining about cyclomatic complexity.